File: Top_Interview_150/Hashmap/python/49.py
from collections import defaultdict
from typing import List


class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        d = defaultdict(list)
        for s in strs:
            d["".join(sorted(s))].append(s)

        return list(d.values())


solution = Solution()
print(solution.groupAnagrams(["eat", "tea", "tan", "ate", "nat", "bat"]))

File: Top_Interview_150/Hashmap/cpp/49.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <algorithm>

using namespace std;

class Solution
{
public:
    vector<vector<string>> groupAnagrams(vector<string> &strs)
    {
        unordered_map<string, vector<string>> d;
        for (string s : strs)
        {
            string sorted_s = s;
            sort(sorted_s.begin(), sorted_s.end());
            d[sorted_s].push_back(s);
        }

        vector<vector<string>> result;
        for (const auto &[key, value] : d)
        {
            result.push_back(value);
        }

        return result;
    }
};